core/src/main/scala/org/apache/spark/launcher/LauncherBackend.scala

```diff
@@ -71,6 +100,9 @@ private[spark] abstract class LauncherBackend {
     if (connection != null && lastState != state) {
       connection.send(new SetState(state))
       lastState = state
+      if (!_isConnected && stopOnShutdown) {
```

Comment:
You haven't addressed my previous comment here: https://github.com/apache/spark/pull/15009#discussion_r84787094

If the backend is not connected to the launcher, it may mean the launcher has explicitly disconnected from it. So the launcher explicitly wants the app to keep running when the connection is closed. This seems to be breaking that.

Can you clarify what you're trying to achieve here? It feels to me like you want the launcher to stop the application when the launcher's JVM exits. And that means handling this in the launcher code, not here.